File: ios-native/NeobankNative/Sources/Features/Kyc/KycModels.swift
import Foundation

enum KycStatus: Sendable {
    case pending
    case approved
    case rejected
    case manualReview

    init(raw: String) {
        switch raw {
        case "approved": self = .approved
        case "rejected": self = .rejected
        case "manual_review": self = .manualReview
        default: self = .pending
        }
    }
}

struct KycStatusInfo: Decodable, Sendable {
    let status: KycStatus
    let rejectionReason: String?

    private enum CodingKeys: String, CodingKey {
        case status, rejectionReason
    }

    init(status: KycStatus, rejectionReason: String?) {
        self.status = status
        self.rejectionReason = rejectionReason
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        status = KycStatus(raw: try container.decode(String.self, forKey: .status))
        rejectionReason = try container.decodeIfPresent(String.self, forKey: .rejectionReason)
    }
}

struct KycSubmitResult: Decodable, Sendable {
    let kycCaseId: String
    let status: KycStatus
    let walletId: String?
    let rejectionReason: String?

    private enum CodingKeys: String, CodingKey {
        case kycCaseId, status, walletId, rejectionReason
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        kycCaseId = try container.decode(String.self, forKey: .kycCaseId)
        status = KycStatus(raw: try container.decode(String.self, forKey: .status))
        walletId = try container.decodeIfPresent(String.self, forKey: .walletId)
        rejectionReason = try container.decodeIfPresent(String.self, forKey: .rejectionReason)
    }
}

File: android-native/app/src/main/kotlin/dev/horobets/neobanknative/features/kyc/KycModels.kt
package dev.horobets.neobanknative.features.kyc

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

enum class KycStatus {
    PENDING,
    APPROVED,
    REJECTED,
    MANUAL_REVIEW,
    ;

    companion object {
        fun fromRaw(raw: String): KycStatus = when (raw) {
            "approved" -> APPROVED
            "rejected" -> REJECTED
            "manual_review" -> MANUAL_REVIEW
            else -> PENDING
        }
    }
}

@Serializable
data class KycStatusInfo(
    @SerialName("status") val rawStatus: String,
    val rejectionReason: String? = null,
) {
    val status: KycStatus get() = KycStatus.fromRaw(rawStatus)
}

@Serializable
data class KycSubmitResult(
    val kycCaseId: String,
    @SerialName("status") val rawStatus: String,
    val walletId: String? = null,
    val rejectionReason: String? = null,
) {
    val status: KycStatus get() = KycStatus.fromRaw(rawStatus)
}
